src/rpc/rpc_payment_signature.cpp

```diff
@@ -0,0 +1,107 @@
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <inttypes.h>
+#include <stdlib.h>
+#include "include_base_utils.h"
+#include "string_tools.h"
+#include "rpc_payment_signature.h"
+
+#undef MONERO_DEFAULT_LOG_CATEGORY
+#define MONERO_DEFAULT_LOG_CATEGORY "daemon.rpc.payment"
+
+#define TIMESTAMP_LEEWAY (60 * 1000000) /* 60 seconds, in microseconds */
+
+namespace cryptonote
+{
+  const boost::posix_time::ptime rpc_payment_epoch(boost::gregorian::date(1970, 1, 1));
+
+  std::string make_rpc_payment_signature(const crypto::secret_key &skey)
+  {
+    std::string s;
+    crypto::public_key pkey;
+    crypto::secret_key_to_public_key(skey, pkey);
+    crypto::signature sig;
+    const uint64_t now = (boost::posix_time::microsec_clock::universal_time() - rpc_payment_epoch).total_microseconds();
+    char ts[17];
+    snprintf(ts, sizeof(ts), "%16.16" PRIx64, now);
+    ts[16] = 0;
+    CHECK_AND_ASSERT_MES(strlen(ts) == 16, "", "Invalid time conversion");
+    crypto::hash hash;
+    crypto::cn_fast_hash(ts, 16, hash);
+    crypto::generate_signature(hash, pkey, skey, sig);
+    s = epee::string_tools::pod_to_hex(pkey) + ts + epee::string_tools::pod_to_hex(sig);
+    return s;
+  }
+
+  bool verify_rpc_payment_signature(const std::string &message, crypto::public_key &pkey, uint64_t &ts)
+  {
+    if (message.size() != 2 * sizeof(crypto::public_key) + 16 + 2 * sizeof(crypto::signature))
+    {
+      MDEBUG("Bad message size: " << message.size());
+      return false;
+    }
+    const std::string pkey_string = message.substr(0, 2 * sizeof(crypto::public_key));
+    const std::string ts_string = message.substr(2 * sizeof(crypto::public_key), 16);
+    const std::string signature_string = message.substr(2 * sizeof(crypto::public_key) + 16);
+    if (!epee::string_tools::hex_to_pod(pkey_string, pkey))
+    {
+      MDEBUG("Bad client id");
+      return false;
+    }
+    crypto::signature signature;
+    if (!epee::string_tools::hex_to_pod(signature_string, signature))
+    {
+      MDEBUG("Bad signature");
+      return false;
+    }
+    crypto::hash hash;
+    crypto::cn_fast_hash(ts_string.data(), 16, hash);
+    if (!crypto::check_signature(hash, pkey, signature))
+    {
+      MDEBUG("signature does not verify");
+      return false;
+    }
+    char *endptr = NULL;
+    errno = 0;
+    unsigned long long ull = strtoull(ts_string.c_str(), &endptr, 16);
+    if (ull == ULLONG_MAX && errno == ERANGE)
+    {
+      MDEBUG("bad timestamp");
+      return false;
+    }
+    ts = ull;
+    const uint64_t now = (boost::posix_time::microsec_clock::universal_time() - rpc_payment_epoch).total_microseconds();
+    if (ts > now + TIMESTAMP_LEEWAY)
```

Comment:
What's the story behind this check making sure that the timestamp is not in the future? Is there a special danger with those?

And isn't there quite a danger of false alarms, caused by clients that have set their timezone wrong in the OS, so that their UTC time comes out 1 hour or more in the future compared to the UTC time that the daemon calculates, which then would trigger this test?

Reply:
A request mst have a timestamp greater (or greater or equal) to the previous request's timestamp. If you allow some insane timestamp, then it will prevent any future request using a correct timestamp. If a client has their timezone wrong, I don't see why that would matter for UTC time, unless the client does some dire determine-UTC-from-local-time-and-wrong-timezone.

Reply:
Understood. I think you are right about timezones; clocks that are simply set wrong "by hand" and not yet corrected back by some following NTP lookup might be a bigger problem. Anyway if surprisingly problems ever crop up this simple test will be easy to modify, so probably no worry.

Reply:
The timestamp is what's signed. Signature must verify to check the sender is the owner of the credits. You want the timestamp to be unique (except for some light calls which are allowed) to avoid replay. Using a nonce means you can't use the system from more than one device (unless you copy the nonce from device to device when you switch).

Reply:
Yes, I think I understand that "signed timestamp" system in general and the problems it solves and the protection that it provides. I was only wondering whether this particular check might be too strict, but that has been cleared now :)

Reply:
Updated in next review - but the second part about replays being temporarily possible isn't true due to the tracking in another class (unless the account is pruned).

Reply:
The LEEWAY system is meant for someone who some manages to spoof someone cannot prevent them from using the system again. It is not intended to prevent replay (as you apparently found out since).